# Francis Yang 10/12/2022
## Pandas Assignment
1. How many rows of data are in the DataFrame?

In [1]:
import pandas as pd
data = pd.read_csv('datasets/ksprojects.csv')
data.shape

(65418, 13)

There are 65,418 rows in this dataframe

2. What are the names and data types of the columns?

In [2]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65418 entries, 0 to 65417
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             65418 non-null  int64  
 1   name           65418 non-null  object 
 2   category       65418 non-null  object 
 3   main_category  65418 non-null  object 
 4   deadline       65418 non-null  object 
 5   goal           65418 non-null  float64
 6   launched       65418 non-null  object 
 7   state          65418 non-null  object 
 8   backers        65418 non-null  int64  
 9   country        65418 non-null  object 
 10  usd_pledged    64673 non-null  float64
 11  currency       65418 non-null  object 
 12  pledged        65418 non-null  float64
dtypes: float64(3), int64(2), object(8)
memory usage: 6.5+ MB


3. Do any of the columns contain null values?

In [3]:
nullcheck = data.isnull()
for col in nullcheck.columns:
    print(nullcheck[col].value_counts())
    print()

False    65418
Name: id, dtype: int64

False    65418
Name: name, dtype: int64

False    65418
Name: category, dtype: int64

False    65418
Name: main_category, dtype: int64

False    65418
Name: deadline, dtype: int64

False    65418
Name: goal, dtype: int64

False    65418
Name: launched, dtype: int64

False    65418
Name: state, dtype: int64

False    65418
Name: backers, dtype: int64

False    65418
Name: country, dtype: int64

False    64673
True       745
Name: usd_pledged, dtype: int64

False    65418
Name: currency, dtype: int64

False    65418
Name: pledged, dtype: int64



There are some null values in the `usd_pledged` column

4. Find all successful documentary projects and sort them by the amount pledged. Print the top 10 highest pledges.


In [4]:
data = data.sort_values('pledged',ascending = False)

5. Create a new column named average_per_backer and set its value to the total amount pledged / number of backers.


In [5]:
data['average_per_backer'] = data['pledged']/data['backers']
data.head()

id                                            name  \
12253  1033978702          OUYA: A New Kind of Video Game Console   
62904  1386523707                   Fidget Cube: A Vinyl Desk Toy   
26917  1152735261  Pono Music - Where Your Soul Rediscovers Music   
45290  1074800880               Bloodstained: Ritual of the Night   
46419  1045500705                         Zombicide: Black Plague   

              category main_category             deadline      goal  \
12253  Gaming Hardware         Games  2012-08-09 07:00:00  950000.0   
62904   Product Design        Design  2016-10-20 03:00:00   15000.0   
26917            Sound    Technology  2014-04-15 19:30:42  800000.0   
45290      Video Games         Games  2015-06-13 05:00:00  500000.0   
46419   Tabletop Games         Games  2015-07-07 02:00:00  125000.0   

                  launched       state  backers country  usd_pledged currency  \
12253  2012-07-10 14:44:41  successful    63416      US    8596480.0      USD   
62904  2016-08-30 22:02:09  successful   154926      US      13770.0      USD   
26917  2014-03-11 18:30:42  successful    18220      US    6225360.0      USD   
45290  2015-05-11 18:58:41  successful    64867      US    5545990.0      USD   
46419  2015-06-08 20:59:38  successful    20915      US    4079200.0      USD   

         pledged  average_per_backer  
12253  8596480.0          135.556957  
62904  6465690.0           41.734054  
26917  6225360.0          341.677278  
45290  5545990.0           85.497865  
46419  4079200.0          195.037055

6. What happened to the rows with 0 backers? How can this be dealt with?


This would cause division by zero. We could add a conditional statement or replace values with Nan

7. Drop all rows with 0 backers then repeat the previous exercise.


In [6]:
data['average_per_backer'].isna().value_counts()

False    56136
True      9282
Name: average_per_backer, dtype: int64

In [9]:
data2 = data.dropna(subset=['average_per_backer'])
data2

id                                            name  \
12253  1033978702          OUYA: A New Kind of Video Game Console   
62904  1386523707                   Fidget Cube: A Vinyl Desk Toy   
26917  1152735261  Pono Music - Where Your Soul Rediscovers Music   
45290  1074800880               Bloodstained: Ritual of the Night   
46419  1045500705                         Zombicide: Black Plague   
...           ...                                             ...   
43036  1387088345                    Will and Dan's Adventure Van   
50633  1049445142                                  Alleanza Comic   
63275  1354388458                          Bamzy Baby Productions   
40169  1237363045                                     StudentList   
6198   1368692455                      The 99% - We Will Be Heard   

              category main_category             deadline      goal  \
12253  Gaming Hardware         Games  2012-08-09 07:00:00  950000.0   
62904   Product Design        Design  2016-10-20 03:00:00   15000.0   
26917            Sound    Technology  2014-04-15 19:30:42  800000.0   
45290      Video Games         Games  2015-06-13 05:00:00  500000.0   
46419   Tabletop Games         Games  2015-07-07 02:00:00  125000.0   
...                ...           ...                  ...       ...   
43036           Places   Photography  2015-04-30 14:21:23    1248.0   
50633           Comics        Comics  2015-10-21 14:21:33     300.0   
63275        Animation  Film & Video  2016-11-01 18:11:00   15000.0   
40169              Web    Technology  2015-03-06 21:47:48    1000.0   
6198      Film & Video  Film & Video  2011-11-01 05:59:00    5000.0   

                  launched       state  backers country   usd_pledged  \
12253  2012-07-10 14:44:41  successful    63416      US  8.596480e+06   
62904  2016-08-30 22:02:09  successful   154926      US  1.377000e+04   
26917  2014-03-11 18:30:42  successful    18220      US  6.225360e+06   
45290  2015-05-11 18:58:41  successful    64867      US  5.545990e+06   
46419  2015-06-08 20:59:38  successful    20915      US  4.079200e+06   
...                    ...         ...      ...     ...           ...   
43036  2015-03-31 14:21:23      failed        1      AU  7.723840e-01   
50633  2015-09-16 14:21:33      failed        1      ES  1.131460e+00   
63275  2016-09-02 20:04:01      failed        1      US  0.000000e+00   
40169  2015-02-04 21:47:48      failed        1      US  1.000000e+00   
6198   2011-10-11 23:14:58      failed        1      US  1.000000e+00   

      currency    pledged  average_per_backer  
12253      USD  8596480.0          135.556957  
62904      USD  6465690.0           41.734054  
26917      USD  6225360.0          341.677278  
45290      USD  5545990.0           85.497865  
46419      USD  4079200.0          195.037055  
...        ...        ...                 ...  
43036      AUD        1.0            1.000000  
50633      EUR        1.0            1.000000  
63275      USD        1.0            1.000000  
40169      USD        1.0            1.000000  
6198       USD        1.0            1.000000  

[56136 rows x 14 columns]

8. Create a DataFrame whose index is the first 10 letters of the alphabet and contains two more columns with first 10 prime numbers and the first 10 fibonacci numbers.


In [14]:
letters = ['a','b','c','d','e','f','g','h','i','j']
prime_nums = [1,3,5,7,9,11,13,17,19,23]
fib_nums = [1,1,2,3,5,8,13,21,34,55]
df = pd.DataFrame({'prime_num':prime_nums,'fib_num':fib_nums},index=letters)
df

prime_num  fib_num
a          1        1
b          3        1
c          5        2
d          7        3
e          9        5
f         11        8
g         13       13
h         17       21
i         19       34
j         23       55

9. Merge two DataFrames into a single DataFrame: nycflights13.

Please use the following dataset for merging purposes: airlines & flights

In [15]:
airlines = pd.read_csv('datasets/airlines.csv')
flights = pd.read_csv('datasets/flights.csv')

In [23]:
airlines.shape
airlines.head()

carrier                    name
0      9E       Endeavor Air Inc.
1      AA  American Airlines Inc.
2      AS    Alaska Airlines Inc.
3      B6         JetBlue Airways
4      DL    Delta Air Lines Inc.

In [25]:
flights.shape
flights.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute            time_hour  
0      1400     5      15  2013-01-01 05:00:00  
1      1416     5      29  2013-01-01 05:00:00  
2      1089     5      40  2013-01-01 05:00:00  
3      1576     5      45  2013-01-01 05:00:00  
4       762     6       0  2013-01-01 06:00:00

The common column is carrier. I will merge airlines into flights because flights is bigger.


In [28]:
nycflights13 = flights.merge(airlines,on='carrier')
nycflights13.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     554.0             558       -4.0     740.0   
3  2013      1    1     558.0             600       -2.0     924.0   
4  2013      1    1     558.0             600       -2.0     923.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             728       12.0      UA    1696  N39463    EWR  ORD     150.0   
3             917        7.0      UA     194  N29129    JFK  LAX     345.0   
4             937      -14.0      UA    1124  N53441    EWR  SFO     361.0   

   distance  hour  minute            time_hour                   name  
0      1400     5      15  2013-01-01 05:00:00  United Air Lines Inc.  
1      1416     5      29  2013-01-01 05:00:00  United Air Lines Inc.  
2       719     5      58  2013-01-01 05:00:00  United Air Lines Inc.  
3      2475     6       0  2013-01-01 06:00:00  United Air Lines Inc.  
4      2565     6       0  2013-01-01 06:00:00  United Air Lines Inc.

10. Are there any missing values in this DataFrame? If so, drop them.

11. Merge these two DataFrames into a single DataFrame called census.

Please use the following dataset for merging purposes: census-divisions & state-populations

12. Re-shape census so that one column contains all population measures, and another the year attributes.

13. Group the data by year and summarize it.

14. Group the data by region, division and year and summarize it.


15. Create a crosstab between state and category showing amount pledged as the value. (ksproject)

In [32]:
temp = [12,15,17,18,23,27,29,32,25,21,20,19]
sales = [2230,3220,3550,3250,5530,7200,8600,9180,6420,4830,4980,4850]
from numpy import corrcoef
corrcoef(temp,sales)

array([[1.        , 0.98413159],
       [0.98413159, 1.        ]])